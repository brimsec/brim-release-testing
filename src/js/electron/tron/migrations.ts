import path from "path"

import {last} from "../../lib/Array"
import lib from "../../lib"

const dir = path.join(__dirname, "../../state/migrations")

type Migration = {
  version: number
  migrate: Function
}

type VersionedData = {
  version: number
  data: any
}

export type Migrations = {
  runPending: (arg0: any) => any
  run: (arg0: VersionedData, arg1: Migration[]) => VersionedData
  getLatestVersion: () => number
  getPending: () => Migration[]
  getAll: () => Migration[]
  setCurrentVersion: (arg0: number) => void
}

export default async function migrations(
  currentVersion: string | number = 0
): Promise<Migrations> {
  let cv = parseInt(currentVersion.toString())
  // $FlowFixMe
  const files = await lib.file(dir).contents()
  const migrations = files
    .filter(excludeTests)
    .map(build)
    .sort((a, b) => a.version - b.version)

  return {
    runPending(state: VersionedData) {
      return this.run(state, this.getPending())
    },

    run(state: VersionedData, migrations: Migration[]) {
      for (const {version, migrate} of migrations) {
        state.data = migrate(state.data)
        state.version = version
      }
      return state
    },

    getLatestVersion() {
      return last(migrations).version || 0
    },

    getPending() {
      return migrations.filter((m: Migration) => m.version > cv)
    },

    getAll() {
      return migrations
    },

    setCurrentVersion(version: number) {
      cv = version
    }
  }
}

function excludeTests(file) {
  return !/\.test\.(ts|js)/.test(file)
}

function build(file): Migration {
  // $FlowFixMe
  const migrate = require(path.join(dir, file)).default
  const [version] = file
    .replace(".js", "")
    .replace(".ts", "")
    .split("_")
  return {
    migrate,
    version: parseInt(version)
  }
}
