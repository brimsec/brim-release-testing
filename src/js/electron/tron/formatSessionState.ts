import {GlobalState} from "../../state/globalReducer"
import {WindowState, WindowsState} from "./windowManager"

export type SessionState = {
  order: string[]
  windows: {
    [key: string]: {
      name: string
      position: [number, number]
      size: [number, number]
      state: Object
    }
  }
  globalState: GlobalState
}

export default function(
  windows: WindowsState,
  globalState: GlobalState
): SessionState {
  const groupById = (all, id) => ({
    ...all,
    [id]: getWindowData(windows[id])
  })

  const order = getWindowOrder(windows)

  return {
    order,
    windows: order.reduce(groupById, {}),
    globalState
  }
}

function getWindowOrder(windows: WindowsState): string[] {
  return Object.entries(windows) // $FlowFixMe
    .sort((a, b) => a[1].lastFocused - b[1].lastFocused)
    .map((e) => e[0])
}

function getWindowData({name, state, size, position}: WindowState) {
  return {name, state, size, position}
}
