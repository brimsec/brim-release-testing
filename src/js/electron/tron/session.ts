import {get} from "lodash"
import log from "electron-log"

import {app} from "electron"
import path from "path"

import {SessionState} from "./formatSessionState"
import {isNumber} from "../../lib/is"
import lib from "../../lib"
import tron from "./"

export default function session(path: string = sessionStateFile()) {
  let version = 0

  return {
    getVersion() {
      return version
    },

    save(data: SessionState, p: string = path) {
      return lib.file(p).write(JSON.stringify({version, data}))
    },

    load: async function(): Promise<SessionState | null | undefined> {
      const migrator = await tron.migrations()
      const file = lib.file(path)

      version = migrator.getLatestVersion()

      if (await file.exists()) {
        return await file
          .read()
          .then(JSON.parse)
          .then((state) => migrate(state, migrator))
          .then((state) => state.data)
          .catch((e) => {
            log.error("Unable to load session state")
            log.error(e)
            return undefined
          })
      } else {
        return undefined
      }
    }
  }
}

function sessionStateFile() {
  // This can't be a const because we adjust the
  // userData path first thing inside main().
  return path.join(app.getPath("userData"), "appState.json")
}

type VersionedState = {version: number; data: SessionState | null | undefined}

async function migrate(appState, migrator): Promise<VersionedState> {
  const state = ensureVersioned(appState)

  if (!canMigrate(state)) {
    log.info("migrations: unsupported version, using fresh state")
    return freshState(migrator.getLatestVersion())
  }

  migrator.setCurrentVersion(state.version)
  const pending = migrator.getPending().length

  log.info(`migrations: currentVersion=${state.version} pending=${pending}`)

  if (pending) {
    try {
      log.info("migrations: running")
      const nextState = migrator.runPending(state)
      log.info(`migrations: currentVersion=${nextState.version}`)
      return nextState
    } catch (e) {
      log.error("Unable to migrate data")
      log.error(e)
      return freshState(migrator.getLatestVersion())
    }
  } else {
    return state
  }
}

function ensureVersioned(state) {
  if (isNumber(state.version)) return state
  else
    return {
      version: 0,
      data: state
    }
}

function canMigrate(state: VersionedState) {
  const legacyVersion = get(state.data, "globalState.version")

  if (!legacyVersion) return true // Already migrated up
  if (legacyVersion === "7") return true // Release right before migration support
  return false // Anything other than above is not migratable
}

function freshState(version): VersionedState {
  return {data: undefined, version}
}
