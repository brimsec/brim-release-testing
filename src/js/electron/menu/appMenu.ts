/* @noflow */

import {app, dialog, shell, MenuItemConstructorOptions} from "electron"
import path from "path"

import {$WindowManager} from "../tron/windowManager"
import {Session} from "../tron"
import config from "../config"
import electronIsDev from "../isDev"
import formatSessionState from "../tron/formatSessionState"
import lib from "../../lib"

export default function appMenu(
  send: Function,
  manager: $WindowManager,
  store: any,
  session: Session,
  platform: string = process.platform
): MenuItemConstructorOptions[] {
  const mac = platform === "darwin"
  const __: MenuItemConstructorOptions = {type: "separator"}

  const newWindow: MenuItemConstructorOptions = {
    label: "New Window",
    accelerator: "CmdOrCtrl+N",
    click: () => manager.openWindow("search", {})
  }

  const exit: MenuItemConstructorOptions = {
    label: "Exit",
    click: () => app.quit()
  }

  const aboutBrim: MenuItemConstructorOptions = {
    label: "About Brim",
    click() {
      manager.openAbout()
    }
  }

  const closeWindow: MenuItemConstructorOptions = {
    label: "Close Window",
    click: () => manager.closeWindow()
  }

  const closeTab: MenuItemConstructorOptions = {
    label: "Close Tab",
    click: () => send("closeTab")
  }

  const preferences: MenuItemConstructorOptions = {
    id: "preferences",
    label: platform === "darwin" ? "Preferences..." : "Settings",
    click: () => manager.openPreferences()
  }

  const resetState: MenuItemConstructorOptions = {
    label: "Reset State",
    click: () => {
      send("resetState")
      lib
        .file(config.windowStateFile())
        .remove()
        .catch(() => {})
    }
  }

  const exportResults: MenuItemConstructorOptions = {
    label: "Export Results as ZNG...",
    click: () => send("exportResults")
  }

  const brimMenu: MenuItemConstructorOptions = {
    label: "Brim",
    submenu: [
      aboutBrim,
      __,
      preferences,
      {role: "services", submenu: []},
      __,
      {role: "hide"},
      {role: "hideOthers"},
      {role: "unhide"},
      __,
      {role: "quit"}
    ]
  }

  function fileSubmenu(): MenuItemConstructorOptions[] {
    if (mac) {
      return [newWindow, __, exportResults, __, closeTab, closeWindow]
    } else {
      return [
        newWindow,
        __,
        exportResults,
        __,
        preferences,
        __,
        closeTab,
        closeWindow,
        exit
      ]
    }
  }

  function editSubmenu(): MenuItemConstructorOptions[] {
    const submenu: MenuItemConstructorOptions[] = [
      {role: "undo"},
      {role: "redo"},
      __,
      {role: "cut"},
      {role: "copy"},
      {role: "paste"},
      {role: "pasteAndMatchStyle"},
      {role: "delete"},
      {role: "selectAll"}
    ]
    if (mac) {
      submenu.push(__, {
        label: "Speech",
        submenu: [{role: "startSpeaking"}, {role: "stopSpeaking"}]
      })
    }
    return submenu
  }

  function windowSubmenu(): MenuItemConstructorOptions[] {
    const submenu = [
      {role: "minimize"} as MenuItemConstructorOptions,
      resetState
    ]
    if (mac) {
      submenu.push({role: "close"}, {role: "minimize"}, {role: "zoom"}, __, {
        role: "front"
      })
    }
    return submenu
  }

  function querySubmenu() {
    return [
      {
        label: "Pin Search",
        accelerator: "CmdOrCtrl+K",
        click: () => send("pinSearch")
      },
      {
        label: "Clear Pins",
        accelerator: "CmdOrCtrl+Shift+K",
        click: () => send("clearPins")
      },
      {
        label: "Focus Search Bar",
        accelerator: "CmdOrCtrl+L",
        click: () => send("focusSearchBar")
      },
      __,
      {
        label: "Back",
        accelerator: "CmdOrCtrl+Left",
        click: () => send("back")
      },
      {
        label: "Forward",
        accelerator: "CmdOrCtrl+Right",
        click: () => send("forward")
      }
    ]
  }

  function viewSubmenu(): MenuItemConstructorOptions[] {
    return [
      {role: "reload"},
      {role: "forceReload"},
      {role: "toggleDevTools"},
      __,
      {role: "resetZoom"},
      {role: "zoomIn"},
      {role: "zoomOut"},
      __,
      {
        label: "Toggle Left Pane",
        accelerator: "CmdOrCtrl+[",
        click: () => send("toggleLeftSidebar")
      },
      {
        label: "Toggle Right Pane",
        accelerator: "CmdOrCtrl+]",
        click: () => send("toggleRightSidebar")
      },
      __,
      {role: "togglefullscreen"}
    ]
  }

  function helpSubmenu() {
    const submenu: MenuItemConstructorOptions[] = [
      {
        label: "ZQL Syntax Docs",
        click() {
          shell.openExternal(
            "https://github.com/brimsec/zq/tree/master/zql/docs"
          )
        }
      },
      {
        label: "Slack Support Channel",
        click() {
          shell.openExternal(
            "https://join.slack.com/t/brimsec/shared_invite/zt-cy34xoxg-hZiTKUT~1KdGjlaBIuUUdg"
          )
        }
      },
      {
        label: "Github Repository",
        click() {
          shell.openExternal("https://github.com/brimsec/brim")
        }
      },
      {
        label: "Submit Issue...",
        click() {
          shell.openExternal(
            "https://github.com/brimsec/brim/wiki/Troubleshooting#opening-an-issue"
          )
        }
      }
    ]

    if (!mac) {
      submenu.push(__, aboutBrim)
    }
    return submenu
  }

  const developerMenu = {
    label: "Developer",
    submenu: [
      {
        label: "Save Session for Testing Migrations",
        async click() {
          await manager.fetchWindowStates()
          const root = app.getAppPath()
          const version = session.getVersion()
          const file = path.join(root, `src/js/test/states/${version}.json`)
          const data = formatSessionState(manager.getState(), store.getState())
          await session.save(data, file)
          dialog.showMessageBox({
            message: `Session has been saved`,
            detail: file
          })
        }
      }
    ]
  }

  const template: MenuItemConstructorOptions[] = [
    {label: "File", submenu: fileSubmenu()},
    {label: "Edit", submenu: editSubmenu()},
    {label: "Query", submenu: querySubmenu()},
    {label: "View", submenu: viewSubmenu()},
    {role: "window", submenu: windowSubmenu()},
    {role: "help", submenu: helpSubmenu()}
  ]
  if (mac) template.unshift(brimMenu)
  if (electronIsDev) template.push(developerMenu)
  return template
}
