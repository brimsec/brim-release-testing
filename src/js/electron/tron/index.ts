import migrations from "./migrations"
import session from "./session"
import window from "./window"
import windowManager from "./windowManager"

export type Session = ReturnType<typeof session>

export default {
  window,
  windowManager,
  session,
  migrations
}
