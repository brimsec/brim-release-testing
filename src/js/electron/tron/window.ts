import {BrowserWindow} from "electron"

import {WindowName} from "./windowManager"

export type WindowParams = {
  size: [number, number]
  position?: [number, number]
  query: Object
  id: string
}

export default function window(name: WindowName, params: WindowParams) {
  switch (name) {
    case "search":
      return searchWindow(params)
    case "about":
      return aboutWindow()
    case "detail":
      return detailWindow(params)
    default:
      throw new Error(`Unknown window name: ${name}`)
  }
}

function searchWindow(params) {
  const {size, position, query, id} = params
  const win = new BrowserWindow({
    titleBarStyle: "hidden",
    resizable: true,
    minWidth: 480,
    minHeight: 100,
    webPreferences: {
      nodeIntegration: true,
      experimentalFeatures: true
    }
  }).on("close", (e: any) => {
    // Close handled by the search renderer
    e.preventDefault()
    e.sender.webContents.send("close")
  })

  if (size) {
    const [width, height] = size
    win.setSize(width, height)
  }
  if (position) {
    const [x, y] = position
    win.setPosition(x, y)
  } else {
    win.center()
  }
  win.loadFile("search.html", {query: {...query, id}})

  return win
}

function aboutWindow() {
  const win = new BrowserWindow({
    resizable: false,
    minimizable: false,
    maximizable: false,
    width: 360,
    height: 360,
    webPreferences: {
      nodeIntegration: true
    }
  })
  win.setMenu(null)
  win.center()
  win.loadFile("about.html")
  return win
}

function detailWindow(params) {
  const {size, position, query, id} = params
  const win = new BrowserWindow({
    resizable: true,
    width: 360,
    height: 360,
    webPreferences: {
      nodeIntegration: true
    }
  })
  if (size) {
    win.setSize(size[0], size[1])
  }
  if (position) {
    win.setPosition(position[0], position[1])
  } else {
    win.center()
  }

  win.loadFile("detail.html", {query: {...query, id}})

  return win
}
