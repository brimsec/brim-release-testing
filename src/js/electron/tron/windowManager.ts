import {BrowserWindow, ipcMain} from "electron"

import {NewTabSearchParams} from "../ipc/windows/messages"
import {SessionState} from "./formatSessionState"
import {WindowParams} from "./window"
import {isBoolean} from "../../lib/is"
import brim from "../../brim"
import ipc from "../ipc"
import sendTo from "../ipc/sendTo"
import tron from "./"

export type WindowName = "search" | "about" | "detail"
export type $WindowManager = ReturnType<typeof windowManager>

export type WindowsState = {
  [key: string]: WindowState
}

export type WindowState = {
  ref: BrowserWindow
  name: string
  size?: [number, number]
  position?: [number, number]
  state?: Object
  lastFocused?: number
}

export default function windowManager() {
  let windows: WindowsState = {}
  let isQuitting = false

  return {
    init(session?: SessionState | null | undefined) {
      if (!session || (session && session.order.length === 0))
        return this.openWindow("search")
      for (const id of session.order) {
        const {name, size, position, state} = session.windows[id]
        this.openWindow(name, {size, position, id})
        this.updateWindow(id, {state})
      }
    },

    isQuitting(val?: boolean | null | undefined) {
      if (isBoolean(val)) isQuitting = val
      else return isQuitting
    },

    getState(): WindowsState {
      const state = {}
      for (const id in windows) {
        const win = windows[id]
        if (win.name === "search") {
          state[id] = win
        }
      }
      return state
    },

    fetchWindowStates() {
      return Promise.all(
        Object.keys(windows).map((id) => {
          return new Promise((resolve) => {
            const win = windows[id]
            const channel = brim.randomHash()
            const timeout = setTimeout(() => resolve([id, undefined]), 5000)

            ipcMain.once(channel, (event, state) => {
              clearTimeout(timeout)
              this.updateWindow(id, {state})
              resolve([id, state])
            })

            win.ref.webContents.send("getState", channel)
          })
        })
      )
    },

    getWindows(): WindowState[] {
      return Object.values(windows)
    },

    count(): number {
      return Object.keys(windows).length
    },

    updateWindow(id: string, data: Partial<WindowState>) {
      windows = {
        ...windows,
        [id]: {...windows[id], ...data}
      }
    },

    getWindow(id: string): WindowState {
      return windows[id]
    },

    openSearchTab(searchParams: NewTabSearchParams) {
      let isNewWin = true
      const existingWin = this.getWindows()
        .sort((a, b) => a.lastFocused - b.lastFocused)
        .find((w) => w.name === "search")
      if (existingWin) {
        isNewWin = false
        sendTo(
          existingWin.ref.webContents,
          ipc.windows.newSearchTab({...searchParams, isNewWin})
        )
        existingWin.ref.focus()
        return
      }

      const {win} = this.openWindow("search", {})
      win.ref.webContents.once("did-finish-load", () => {
        sendTo(
          win.ref.webContents,
          ipc.windows.newSearchTab({...searchParams, isNewWin})
        )
      })
    },

    openWindow(name: WindowName, winParams: Partial<WindowParams> = {}) {
      const params = defaultWindowParams(winParams)
      const id = params.id

      const ref = tron
        .window(name, params)
        .on("focus", () => {
          if (!isQuitting) windows[id].lastFocused = new Date().getTime()
        })
        .on("closed", () => {
          if (!isQuitting) delete windows[id]
        })

      const win = {ref, name, lastFocused: new Date().getTime()}
      windows[id] = win

      return {id, win}
    },

    openAbout() {
      const about = this.getWindows().find((w) => w.name === "about")
      if (about) {
        about.ref.focus()
      } else {
        this.openWindow("about")
      }
    },

    openPreferences() {
      const win = this.getWindows()
        .sort((a, b) => (b.lastFocused || 0) - (a.lastFocused || 0))
        .find((w) => w.name === "search")

      if (win) {
        win.ref.webContents.send("showPreferences")
      } else {
        const {win} = this.openWindow("search", {})

        win.ref.webContents.once("did-finish-load", () => {
          win.ref.webContents.send("showPreferences")
        })
      }
    },

    closeWindow() {
      const win = BrowserWindow.getFocusedWindow()
      if (win) win.close()
    },

    destroyWindow(win: BrowserWindow) {
      if (win) win.destroy()
    }
  }
}

function defaultWindowParams(params: Partial<WindowParams>): WindowParams {
  return {
    size: [1000, 800],
    id: brim.randomHash(),
    query: {},
    ...params
  }
}
