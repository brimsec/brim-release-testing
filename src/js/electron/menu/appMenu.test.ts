import appMenu from "./appMenu"
import initTestStore from "../../test/initTestStore"
import tron from "../tron"
import {$WindowManager} from "../tron/windowManager"

const mockSend = jest.fn()
const mockWindownManager = {} as $WindowManager
const mockStore = initTestStore()

test("app menu mac", async () => {
  const mockSession = await tron.session()
  const menu = appMenu(
    mockSend,
    mockWindownManager,
    mockStore,
    mockSession,
    "darwin"
  )

  expect(menu).toMatchSnapshot()
})

test("app menu windows", async () => {
  const mockSession = await tron.session()

  const menu = appMenu(
    mockSend,
    mockWindownManager,
    mockStore,
    mockSession,
    "win32"
  )
  expect(menu).toMatchSnapshot()
})
