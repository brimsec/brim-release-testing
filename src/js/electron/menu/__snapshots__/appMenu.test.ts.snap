// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app menu mac 1`] = `
Array [
  Object {
    "label": "Brim",
    "submenu": Array [
      Object {
        "click": [Function],
        "label": "About Brim",
      },
      Object {
        "type": "separator",
      },
      Object {
        "click": [Function],
        "id": "preferences",
        "label": "Preferences...",
      },
      Object {
        "role": "services",
        "submenu": Array [],
      },
      Object {
        "type": "separator",
      },
      Object {
        "role": "hide",
      },
      Object {
        "role": "hideOthers",
      },
      Object {
        "role": "unhide",
      },
      Object {
        "type": "separator",
      },
      Object {
        "role": "quit",
      },
    ],
  },
  Object {
    "label": "File",
    "submenu": Array [
      Object {
        "accelerator": "CmdOrCtrl+N",
        "click": [Function],
        "label": "New Window",
      },
      Object {
        "type": "separator",
      },
      Object {
        "click": [Function],
        "label": "Export Results as ZNG...",
      },
      Object {
        "type": "separator",
      },
      Object {
        "click": [Function],
        "label": "Close Tab",
      },
      Object {
        "click": [Function],
        "label": "Close Window",
      },
    ],
  },
  Object {
    "label": "Edit",
    "submenu": Array [
      Object {
        "role": "undo",
      },
      Object {
        "role": "redo",
      },
      Object {
        "type": "separator",
      },
      Object {
        "role": "cut",
      },
      Object {
        "role": "copy",
      },
      Object {
        "role": "paste",
      },
      Object {
        "role": "pasteAndMatchStyle",
      },
      Object {
        "role": "delete",
      },
      Object {
        "role": "selectAll",
      },
      Object {
        "type": "separator",
      },
      Object {
        "label": "Speech",
        "submenu": Array [
          Object {
            "role": "startSpeaking",
          },
          Object {
            "role": "stopSpeaking",
          },
        ],
      },
    ],
  },
  Object {
    "label": "Query",
    "submenu": Array [
      Object {
        "accelerator": "CmdOrCtrl+K",
        "click": [Function],
        "label": "Pin Search",
      },
      Object {
        "accelerator": "CmdOrCtrl+Shift+K",
        "click": [Function],
        "label": "Clear Pins",
      },
      Object {
        "accelerator": "CmdOrCtrl+L",
        "click": [Function],
        "label": "Focus Search Bar",
      },
      Object {
        "type": "separator",
      },
      Object {
        "accelerator": "CmdOrCtrl+Left",
        "click": [Function],
        "label": "Back",
      },
      Object {
        "accelerator": "CmdOrCtrl+Right",
        "click": [Function],
        "label": "Forward",
      },
    ],
  },
  Object {
    "label": "View",
    "submenu": Array [
      Object {
        "role": "reload",
      },
      Object {
        "role": "forceReload",
      },
      Object {
        "role": "toggleDevTools",
      },
      Object {
        "type": "separator",
      },
      Object {
        "role": "resetZoom",
      },
      Object {
        "role": "zoomIn",
      },
      Object {
        "role": "zoomOut",
      },
      Object {
        "type": "separator",
      },
      Object {
        "accelerator": "CmdOrCtrl+[",
        "click": [Function],
        "label": "Toggle Left Pane",
      },
      Object {
        "accelerator": "CmdOrCtrl+]",
        "click": [Function],
        "label": "Toggle Right Pane",
      },
      Object {
        "type": "separator",
      },
      Object {
        "role": "togglefullscreen",
      },
    ],
  },
  Object {
    "role": "window",
    "submenu": Array [
      Object {
        "role": "minimize",
      },
      Object {
        "click": [Function],
        "label": "Reset State",
      },
      Object {
        "role": "close",
      },
      Object {
        "role": "minimize",
      },
      Object {
        "role": "zoom",
      },
      Object {
        "type": "separator",
      },
      Object {
        "role": "front",
      },
    ],
  },
  Object {
    "role": "help",
    "submenu": Array [
      Object {
        "click": [Function],
        "label": "ZQL Syntax Docs",
      },
      Object {
        "click": [Function],
        "label": "Slack Support Channel",
      },
      Object {
        "click": [Function],
        "label": "Github Repository",
      },
      Object {
        "click": [Function],
        "label": "Submit Issue...",
      },
    ],
  },
]
`;

exports[`app menu windows 1`] = `
Array [
  Object {
    "label": "File",
    "submenu": Array [
      Object {
        "accelerator": "CmdOrCtrl+N",
        "click": [Function],
        "label": "New Window",
      },
      Object {
        "type": "separator",
      },
      Object {
        "click": [Function],
        "label": "Export Results as ZNG...",
      },
      Object {
        "type": "separator",
      },
      Object {
        "click": [Function],
        "id": "preferences",
        "label": "Settings",
      },
      Object {
        "type": "separator",
      },
      Object {
        "click": [Function],
        "label": "Close Tab",
      },
      Object {
        "click": [Function],
        "label": "Close Window",
      },
      Object {
        "click": [Function],
        "label": "Exit",
      },
    ],
  },
  Object {
    "label": "Edit",
    "submenu": Array [
      Object {
        "role": "undo",
      },
      Object {
        "role": "redo",
      },
      Object {
        "type": "separator",
      },
      Object {
        "role": "cut",
      },
      Object {
        "role": "copy",
      },
      Object {
        "role": "paste",
      },
      Object {
        "role": "pasteAndMatchStyle",
      },
      Object {
        "role": "delete",
      },
      Object {
        "role": "selectAll",
      },
    ],
  },
  Object {
    "label": "Query",
    "submenu": Array [
      Object {
        "accelerator": "CmdOrCtrl+K",
        "click": [Function],
        "label": "Pin Search",
      },
      Object {
        "accelerator": "CmdOrCtrl+Shift+K",
        "click": [Function],
        "label": "Clear Pins",
      },
      Object {
        "accelerator": "CmdOrCtrl+L",
        "click": [Function],
        "label": "Focus Search Bar",
      },
      Object {
        "type": "separator",
      },
      Object {
        "accelerator": "CmdOrCtrl+Left",
        "click": [Function],
        "label": "Back",
      },
      Object {
        "accelerator": "CmdOrCtrl+Right",
        "click": [Function],
        "label": "Forward",
      },
    ],
  },
  Object {
    "label": "View",
    "submenu": Array [
      Object {
        "role": "reload",
      },
      Object {
        "role": "forceReload",
      },
      Object {
        "role": "toggleDevTools",
      },
      Object {
        "type": "separator",
      },
      Object {
        "role": "resetZoom",
      },
      Object {
        "role": "zoomIn",
      },
      Object {
        "role": "zoomOut",
      },
      Object {
        "type": "separator",
      },
      Object {
        "accelerator": "CmdOrCtrl+[",
        "click": [Function],
        "label": "Toggle Left Pane",
      },
      Object {
        "accelerator": "CmdOrCtrl+]",
        "click": [Function],
        "label": "Toggle Right Pane",
      },
      Object {
        "type": "separator",
      },
      Object {
        "role": "togglefullscreen",
      },
    ],
  },
  Object {
    "role": "window",
    "submenu": Array [
      Object {
        "role": "minimize",
      },
      Object {
        "click": [Function],
        "label": "Reset State",
      },
    ],
  },
  Object {
    "role": "help",
    "submenu": Array [
      Object {
        "click": [Function],
        "label": "ZQL Syntax Docs",
      },
      Object {
        "click": [Function],
        "label": "Slack Support Channel",
      },
      Object {
        "click": [Function],
        "label": "Github Repository",
      },
      Object {
        "click": [Function],
        "label": "Submit Issue...",
      },
      Object {
        "type": "separator",
      },
      Object {
        "click": [Function],
        "label": "About Brim",
      },
    ],
  },
]
`;
